experimental/framework/fluid-static/src/signalManager.ts

```diff
@@ -0,0 +1,178 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { EventEmitter } from "events";
+import {
+    DataObject,
+    DataObjectFactory,
+} from "@fluidframework/aqueduct";
+import { IErrorEvent } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { Jsonable } from "@fluidframework/datastore-definitions";
+import { IInboundSignalMessage } from "@fluidframework/runtime-definitions";
+
+// TODO:
+// add way to mark with current sequence number for ordering signals relative to ops
+// async listener support
+
+export type SignalListener = (clientId: string, local: boolean, payload: Jsonable) => void;
+
+export interface ISignalManager {
+    /**
+     * Adds a listener for the specified signal at the end of the listeners array.  Multiple calls
+     * with the same signal name and listener will add it (and call it) multiple times.  Listeners
+     * are called in the order they are added.
+     * @param signalName - The name of the signal
+     * @param listener - The callback signal handler
+     * @returns This ISignalManager
+     */
+    registerListener(signalName: string, listener: SignalListener): ISignalManager;
+
+    /**
+     * Remove the last added instance of the listener for the specified signal in the listeners
+     * array if present.  If a listener has been added multiple times, it must be removed that many
+     * times as well to remove all instances.
+     *
+     * TODO: specify behavior around in-flight calls/mutability
+     * @param signalName - The name of the signal
+     * @param listener - The callback signal handler
+     * @returns This ISignalManager
+     */
+    deregisterListener(signalName: string, listener: SignalListener): ISignalManager;
+
+    submitSignal(signalName: string, payload?: Jsonable);
+
+    registerBroadcastListener(signalName: string, listener: SignalListener): ISignalManager;
+
+    deregisterBroadcastListener(signalName: string, listener: SignalListener): ISignalManager;
+
+    requestBroadcast(signalName: string, payload?: Jsonable);
+}
+
+/**
+ *
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types
+export class SignalManager extends DataObject<{}, undefined, IErrorEvent> implements EventEmitter, ISignalManager {
```

Comment:
i wouldn't make this a DataObject, i would just have it be a class that can be instantiated over anything that support submitting signals.

Reply:
Helio and I chatted about this for a bit and, although I was also opposed at first, there are two things this provides that I really like.

First, is that it binds the signal eventing to an object and not an id. This means when a developer creates a new SignalManager they will only see and get signals within the scope of this manager.

Second, it doesn't introduce new concept to the 3P API. Since DataObjects are already first class citizens this would simply be created and managed the same way DataObjects currently are. We don't need to introduce a signal API on the FluidContainer or any additional concepts specific to signals.

Reply:
Latest iteration shows it as a regular class.  It might be more "proper" this way, but like Skyler mentions this complicates its usage with fluid-static and will require FluidContainer changes because the v1 there isn't supporting custom DOs yet.

Reply:
@anthony-murphy @ChumpChief Circling back on this, I still think providing this as a DO is the best approach for the 3P public preview simply because custom DOs are not part of the supported feature set there, and we want to enable presence/audience scenarios that might want signals.  If it weren't for the custom DO restriction I'd be completely for the standalone.  The alternative would be adding something signal-specific on the FluidContainer API, which undesirably bloats it and runs counter to the idea of modular pieces and functionality we're trying to promote with v1.  It could also be provided as the standalone class wrapping a signaler along with a DO version that wraps the standalone class, but I'm not sure that's better than just having a DO.

Reply:
I'm not sure what I'm missing. we currently support two kinds of signals, data store and container runtime. data store are routed to the data store, runtime are routed to runtime respectively. It seems a bit redundant to create a data store, and a map which both have data level side effects for every signal manager. Is there an example of how this fits into 3p? as focus tracker should already have access to signals directly as it is a data object

Reply:
The way I created FocusTracker is probably distracting from the actual 3p scenario.  I created FocusTracker first before SignalManager as a ServiceAudience example/exploration - it needed to be a DO to use signals.  The problem then is that this isn't actually supported in the plan for 3p public preview (we're not supporting creating a custom DO), and I should have completely unwrapped the DO aspect of it as part of the first draft integration with the DO version of SignalManager.

Agree on the redundancy.  Putting it on the FluidContainer would address that and keep it as a standalone class...

Reply:
also, don't consider me block here. i just saw something suspicious and wanted to call it out. we had a real problem in the past where partners were creating data stores for every service they had, and this led to all kinds of issues with file size, and performance. It was the reason we introduced the container service pattern, which are stateless. just want to make sure were thinking through the patterns we use here, and the implications of those patterns.